src/azure-cli/azure/cli/command_modules/synapse/operations/accesscontrol.py

```diff
@@ -61,32 +73,96 @@ def delete_role_assignment(cmd, workspace_name, ids=None, assignee=None, role=No
+                                        .format(assignee)))
+        if not result:
+            result = list(client.service_principals.list(filter="displayName eq '{}'".format(assignee)))
+        if not result:
+            result = list(client.groups.list(filter="mail eq '{}'".format(assignee)))
+        if not result and is_guid(assignee):  # assume an object id, let us verify it
+            result = _get_object_stubs(client, [assignee])
+
+        # 2+ matches should never happen, so we only check 'no match' here
+        if not result:
+            raise CLIError("Cannot find user or group or service principal in graph database for '{assignee}'. "
+                           "If the assignee is a principal id, make sure the corresponding principal is created "
+                           "with 'az ad sp create --id {assignee}'.".format(assignee=assignee))
+
+        return result[0].object_id
+    except (CloudError, GraphErrorException):
+        if fallback_to_object_id and is_guid(assignee):
+            return assignee
+        raise
+
+
+def _get_object_stubs(graph_client, assignees):
+    from azure.graphrbac.models import GetObjectsParameters
+    result = []
+    assignees = list(assignees)  # callers could pass in a set
+    for i in range(0, len(assignees), 1000):
+        params = GetObjectsParameters(include_directory_object_references=True, object_ids=assignees[i:i + 1000])
+        result += list(graph_client.objects.get_objects_by_object_ids(params))
+    return result
+
+
+def _error_caused_by_role_assignment_exists(ex):
+    return getattr(ex, 'status_code', None) == 409 and 'role assignment already exists' in ex.message
+
+
+def _create_role_assignment(cmd, workspace_name, role, assignee, scope=None, item=None, item_type=None,
+                            resolve_assignee=True, assignee_principal_type=None, assignment_id=None):
+    """Prepare scope, role ID and resolve object ID from Graph API."""
+    scope = _build_role_scope(workspace_name, scope, item, item_type)
     role_id = _resolve_role_id(cmd, role, workspace_name)
-    client = cf_synapse_client_accesscontrol_factory(cmd.cli_ctx, workspace_name)
-    return client.get_role_definition_by_id(role_id)
+    object_id = _resolve_object_id(cmd, assignee, fallback_to_object_id=True) if resolve_assignee else assignee
+
+    assignment_client = cf_synapse_role_assignments(cmd.cli_ctx, workspace_name)
+    return assignment_client.create_role_assignment(assignment_id if assignment_id is not None else _gen_guid(),
+                                                    role_id, object_id, scope, assignee_principal_type)
+
+
+def _build_role_scope(workspace_name, scope, item, item_type):
+    if scope:
+        return scope
+
+    if item and item_type:
+        # workspaces/{workspaceName}/bigDataPools/{bigDataPoolName}
+        scope = "workspaces/" + workspace_name + "/" + item_type + "/" + item
+    else:
+        scope = "workspaces/" + workspace_name
```

Comment:
I just want to mention that these two may be not equal from my memory:  1.scope = "workspaces/" + workspace_name 2. scope = null

Reply:
Hi @zesluo  As far as my understand, the default scope is workspaces/{workspace_name}, we can discuss and confirm with @idear1203 .

Thanks for throwing it out.